Do not allow snippets of type method or field before a class statement

turkeylurkey opened the pull request:
The issue is that `rest_get` and `mpirc` snippets are provided as suggestions when the user's cursor is between the `package` statement and the `class` statement. This is incorrect and it is easy for a person to stumble across.

This fix helps with this problem but it means these two snippets do not appear when the cursor is just before an `enum` or an inner class statement. This seems like an improvement. 

Is this change acceptable?

Fixes #402

angelozerr commented:
I understand it is an annoying problem but we need to see if in JDT with vscode microprofile we have the same problem.

@datho7561 what do you think about this change?

angelozerr commented:
I will test too in IJ Quarkus if we have the same problem and if we can do something on IJ side.

datho7561 commented:
This issue with this change is this case:

```
package com.github.datho7561;

/**
 * My documentation
 */
public class MyObject {

	public int getValue() {
		return 12;
	}

	|

	static public class MyInnerClass {

	}
}
```

The snippet `mpirc` (for creating a rest client) should be available here, but it isn't. Having a descriptor `BEFORE_CLASS` is useful for annotations that can only be applied to classes (and interfaces and records and such). However, in order to properly handle completion, we need to distinguish between `BEFORE_INNER_CLASS` and `BEFORE_TOP_LEVEL_CLASS`. Currently, in the JDT code, there are two locations where we return `BEFORE_CLASS`; one of them should be changed to `BEFORE_TOP_LEVEL_CLASS` and the other to `BEFORE_INNER_CLASS`. Then we can use this more specific information to better inform snippet completion results.

angelozerr commented:
Thanks  @datho7561 for your relevant information. IJ need to improve your usecase.

turkeylurkey commented:
I removed the previous suggestion since it was only a partial solution. I integrated the information from @datho7561 above and implemented a proper solution using the constants `BEFORE_INNER_CLASS` and `BEFORE_TOP_LEVEL_CLASS`.

I tested this language server jar using my IntelliJ plugin and also the `vscode-microprofile` plugin. The field and method snippets are no longer offered between the package statement and the class statement.

@angelozerr  and @datho7561 , could you please review this fix and let me know?

Next we will need a release of lsp4mp and vscode-microprofile. Thank you for your attention.

angelozerr commented:
@datho7561 is there any chance that you could review the PR please?

datho7561 commented:
Oops just testing the tests to make sure they are passing (or that they aren't failing any worse than before). Didn't realize CI wasn't running.

datho7561 commented:
I ran the tests locally and none of the failures seem related to this PR, so I'll merge.

turkeylurkey commented:
Thanks, folks.
